src/lib.rs

```diff
@@ -0,0 +1,107 @@
+use std::env;
+
+use log::info;
+use tokio::sync::mpsc;
+
+use trin_core::cli::TrinConfig;
+use trin_core::jsonrpc::launch_jsonrpc_server;
+use trin_core::portalnet::protocol::{
+    HistoryNetworkEndpoint, JsonRpcHandler, PortalEndpoint, PortalnetConfig, PortalnetProtocol,
+    StateNetworkEndpoint,
+};
+use trin_history::HistoryRequestHandler;
+use trin_state::StateRequestHandler;
+
+pub async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    println!("Launching trin");
+
+    let trin_config = TrinConfig::new();
+    trin_config.display_config();
+
+    let infura_project_id = match env::var("TRIN_INFURA_PROJECT_ID") {
+        Ok(val) => val,
+        Err(_) => panic!(
+            "Must supply Infura key as environment variable, like:\n\
+            TRIN_INFURA_PROJECT_ID=\"your-key-here\" trin"
+        ),
+    };
+
+    let bootnode_enrs = trin_config
+        .bootnodes
+        .iter()
+        .map(|nodestr| nodestr.parse().unwrap())
+        .collect();
+
+    let portalnet_config = PortalnetConfig {
+        external_addr: trin_config.external_addr,
+        private_key: trin_config.private_key.clone(),
+        listen_port: trin_config.discovery_port,
+        bootnode_enrs,
+        ..Default::default()
+    };
+
+    let mut state_tx = None;
+    let mut history_tx = None;
+    let mut state_handler: Option<StateRequestHandler> = None;
+    let mut history_handler: Option<HistoryRequestHandler> = None;
+
+    if trin_config.networks.iter().any(|val| val == "history") {
```

Comment:
I did like this idea, but in the end I prefer something more along the lines of what @jacobkaufmann suggested [here](https://github.com/ethereum/trin/pull/89#discussion_r702011735) since it gets rid of initializing empty mutable variables. I wasn't able to devise a way to combine both of your refactors (aka a more explicit iteration and no dead variable instantiation) but if you see something I missed lmk.